### PR TITLE
fix: Return 0 for `n_unique()` in group-by context when group is empty

### LINKED ISSUE #15284: `n_unique` returns `null` instead of 0 inside `group_by`
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pl.DataFrame([{"a": i, "b": i % 2 == 0, "c": i % 4} for i in range(10)])
df.group_by("c").agg(
    pl.col("a").filter("b").n_unique()
).sort(by="c")
```

This returns:
```
  c | a 
--- | ---
i64 | u32
  0 | 3
  1 | null
  2 | 2
  3 | null
```

### Log output

_No response_

### Issue description

When the `group_by` and `filter` operation create an empty group together, counting the unique values (`n_unique`) results in `null` instead of `0`.

### Expected behavior

Instead of the null's I would expect 0. The filter empties the list and counting the unique elements of an empty list returns 0.

One can use `fill_null` to get back to 0, but I would have expected this to be the default behavior. I'm also unsure about potential consequences if `b` can be null itself and how that behaves with the filter. I would expect that a null result is appropriate for that.

If `a` is unique one can also get around by using sum:
```
a.group_by("c").agg(
    pl.sum("b")
).sort(by="c")
```

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.10
Index type:           UInt32
Platform:             Linux-6.6.16-linuxkit-x86_64-with-glibc2.35
Python:               3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           0.3.2
deltalake:            <not installed>
fsspec:               2022.10.0
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.3
numpy:                1.22.4
openpyxl:             <not installed>
pandas:               1.5.3
pyarrow:              9.0.0
pydantic:             1.9.0
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           1.4.32
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

Also ran `pip install polars --upgrade and it persisted with version `0.20.16`.
</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #15284.

Series currently return 0 for `n_unique` on empty sets but not in a `group_by` context. This reconciles the two and returns 0 for empty sets.

### CODE REVIEW

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 1]
```diff
@@ -94,7 +94,7 @@ impl Series {
                 _agg_helper_slice::<IdxType, _>(groups, |[first, len]| {
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:94]: Is that meant for `ChunkedArrays` with no null values, or for when the output will have no nulls?

[the remark this answers is not in the document]
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:94]: Thanks for the tips @nameexhaustion!

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 2]
```diff
@@ -17,27 +17,29 @@ impl Series {
...
-                    None
-                } else if !self.has_validity() {
-                    Some(idx.len() as IdxSize)
-                } else {
-                    let take = unsafe { self.take_slice_unchecked(idx) };
-                    Some((take.len() - take.null_count()) as IdxSize)
-                }
-            }),
+            GroupsProxy::Idx(groups) => {
+                agg_helper_idx_on_all_no_null::<IdxType, _>(groups, |idx| {
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:21]: Thanks @nameexhaustion. I'll make a followup PR with that fix.

[the remark this answers is not in the document]
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs]: @nameexhaustion reverted `valid_count` inadvertent change and applied to `n_unique`. I think it's good now.

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**